File: src/main/java/org/javiermf/features/models/Feature.java
package org.javiermf.features.models;


import com.fasterxml.jackson.annotation.JsonIgnore;

public class Feature {

    String name;
    String description;

    @JsonIgnore
    Product product;

    public static Feature withName(Product product, String featureName) {
        Feature feature = new Feature();
        feature.name = featureName;
        feature.product = product;
        return feature;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public Product getProduct() {
        return product;
    }

    public void setProduct(Product product) {
        this.product = product;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Feature)) return false;

        Feature feature = (Feature) o;

        if (!name.equals(feature.name)) return false;
        return product.equals(feature.product);

    }

    @Override
    public int hashCode() {
        int result = name.hashCode();
        result = 31 * result + product.hashCode();
        return result;
    }
}

File: src/main/kotlin/org/javiermf/features/models/Feature.kt
package org.javiermf.features.models


import com.fasterxml.jackson.annotation.JsonIgnore

import javax.persistence.*

@Entity
class Feature {

    @Id
    @GeneratedValue
    var id: Long
        internal set

    @Column(nullable = false)
    var name: String

    @Column
    var description: String

    @JsonIgnore
    @ManyToOne
    var product: Product

    @JsonIgnore
    @ManyToMany(mappedBy = "activedFeatures")
    internal var inConfigurations: Set<ProductConfiguration>? = null

    constructor(id: Long, name: String, description: String, product: Product) {
        this.id = id
        this.name = name
        this.description = description
        this.product = product
    }


    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o !is Feature) return false

        val feature = o as Feature?

        return if (name != feature!!.name) false else product == feature.product

    }

    override fun hashCode(): Int {
        var result = name.hashCode()
        result = 31 * result + product.hashCode()
        return result
    }

    companion object {

        fun withName(product: Product, featureName: String): Feature {
            return Feature(1, featureName, featureName, product)
        }
    }
}
